# Sample EDA

In [1]:
import dask.dataframe as dd
import pandas as pd
pd.set_option('display.max_columns', 500)
from dask_ml.linear_model import LogisticRegression

Since the dataset doesn't have a header we will use the dataset README to create a preliminary header

In [2]:
column_names = ['label','interger1','interger2','interger3','interger4','interger5','interger6',
                'interger7','interger8','interger9','interger10','interger11','interger12','interger13',
                'categorical1','categorical2','categorical3','categorical4','categorical5','categorical6',
                'categorical7','categorical8','categorical9','categorical10','categorical11','categorical12',
                'categorical13','categorical14','categorical15','categorical16','categorical17','categorical18',
                'categorical19','categorical20','categorical21','categorical22','categorical23','categorical24',
                'categorical25','categorical26']

In [3]:
df = dd.read_csv('/home/bahbbc/workspace/display-ads-challenge/dac/train.txt', sep='\t', names = column_names)

In [4]:
answer = dd.read_csv('/home/bahbbc/workspace/display-ads-challenge/dac/test.txt', sep='\t', names=column_names[1:])

In [5]:
answer.head()

interger1  interger2  interger3  interger4  interger5  interger6  \
0        NaN         29       50.0        5.0     7260.0      437.0   
1       27.0         17       45.0       28.0        2.0       28.0   
2        1.0          1       19.0        7.0        1.0        3.0   
3        4.0          1        NaN        6.0     1051.0      134.0   
4        7.0          1       25.0       10.0      139.0       74.0   

   interger7  interger8  interger9  interger10  interger11  interger12  \
0        1.0        4.0       14.0         NaN         1.0         0.0   
1       27.0       29.0       28.0         1.0         1.0         NaN   
2        1.0        7.0        7.0         1.0         1.0         NaN   
3        4.0       35.0       72.0         1.0         1.0         NaN   
4       48.0       13.0       44.0         1.0         8.0         2.0   

   interger13 categorical1 categorical2 categorical3 categorical4  \
0         6.0     5a9ed9b0     a0e12995     a1e14474     08a40877   
1        23.0     68fd1e64     960c983b     9fbfbfd5     38c11726   
2         2.0     09ca0b81     8947f767     a87e61f7     c4ba2a67   
3         6.0     05db9164     532da141     a7ded28e     456b4d8c   
4        12.0     05db9164     207b2d81     2b280564     ad5ffc6b   

  categorical5 categorical6 categorical7 categorical8 categorical9  \
0     25c83c98          NaN     964d1fdd     5b392875     a73ee510   
1     25c83c98     7e0ccccf     fe06fd10     062b5529     a73ee510   
2     25c83c98     7e0ccccf     ce6020cc     062b5529     a73ee510   
3     25c83c98     fbad5c96     5f29da0e     0b153874     a73ee510   
4     25c83c98     7e0ccccf     103c17bc     0b153874     a73ee510   

  categorical10 categorical11 categorical12 categorical13 categorical14  \
0      de89c3d2      59cd5ae7      8d98db20      8b216f7b      1adce6ef   
1      ca53fc84      67360210      895d8bbb      4f8e2224      f862f261   
2      b04d3cfe      70dcd184      899eb56b      aca22cf9      b28479f6   
3      4b344a42      0ad37b4b      8ea37200      f9d99d81      cfef1c29   
4      8e54038a      e6e0c2dc      2a064dba      e9332a03      07d13a8f   

  categorical15 categorical16 categorical17 categorical18 categorical19  \
0      78c64a1d      3ecdadf7      3486227d      1616f155      21ddcdc9   
1      b4cc2435      4c0041e5      e5ba7672      b4abdd09      21ddcdc9   
2      a473257f      88f592e4      d4bb7bd8      bd17c3da      1d04f4a4   
3      abd8f51e      9a9902d0      07c540c4      bdc06043           NaN   
4      0c67c4ca      7d9b60c8      27c07bd6      395856b0      21ddcdc9   

  categorical20 categorical21 categorical22 categorical23 categorical24  \
0      5840adea      2c277e62           NaN      423fab69      54c91918   
1      5840adea      36a7ab86           NaN      32c7478e      85e4d73f   
2      a458ea53      82bdc0bb           NaN      32c7478e      5bdcd9c4   
3           NaN      6dfd157c      ad3062eb      423fab69      ef089725   
4      a458ea53      9c3eb598      ad3062eb      3a171ecb      c0b8dfd6   

  categorical25 categorical26  
0      9b3e8820      e75c9ae9  
1      010f6491      ee63dd9b  
2      010f6491      cca57dcc  
3           NaN           NaN  
4      001f3601      7a2fb9af

In [6]:
df.head()

label  interger1  interger2  interger3  interger4  interger5  interger6  \
0      0        1.0          1        5.0        0.0     1382.0        4.0   
1      0        2.0          0       44.0        1.0      102.0        8.0   
2      0        2.0          0        1.0       14.0      767.0       89.0   
3      0        NaN        893        NaN        NaN     4392.0        NaN   
4      0        3.0         -1        NaN        0.0        2.0        0.0   

   interger7  interger8  interger9  interger10  interger11  interger12  \
0       15.0        2.0      181.0         1.0         2.0         NaN   
1        2.0        2.0        4.0         1.0         1.0         NaN   
2        4.0        2.0      245.0         1.0         3.0         3.0   
3        0.0        0.0        0.0         NaN         0.0         NaN   
4        3.0        0.0        0.0         1.0         1.0         NaN   

   interger13 categorical1 categorical2 categorical3 categorical4  \
0         2.0     68fd1e64     80e26c9b     fb936136     7b4723c4   
1         4.0     68fd1e64     f0cf0024     6f67f7e5     41274cd7   
2        45.0     287e684f     0a519c5c     02cf9876     c18be181   
3         NaN     68fd1e64     2c16a946     a9a87e68     2e17d6f6   
4         0.0     8cf07265     ae46a29d     c81688bb     f922efad   

  categorical5 categorical6 categorical7 categorical8 categorical9  \
0     25c83c98     7e0ccccf     de7995b8     1f89b562     a73ee510   
1     25c83c98     fe6b92e5     922afcc0     0b153874     a73ee510   
2     25c83c98     7e0ccccf     c78204a1     0b153874     a73ee510   
3     25c83c98     fe6b92e5     2e8a689b     0b153874     a73ee510   
4     25c83c98     13718bbd     ad9fa255     0b153874     a73ee510   

  categorical10 categorical11 categorical12 categorical13 categorical14  \
0      a8cd5504      b2cb9c98      37c9c164      2824a5f6      1adce6ef   
1      2b53e5fb      4f1b46f3      623049e6      d7020589      b28479f6   
2      3b08e48b      5f5e6091      8fe001f4      aa655a2f      07d13a8f   
3      efea433b      e51ddf94      a30567ca      3516f6e6      07d13a8f   
4      5282c137      e5d8af57      66a76a26      f06c53ac      1adce6ef   

  categorical15 categorical16 categorical17 categorical18 categorical19  \
0      8ba8b39a      891b62e7      e5ba7672      f54016b9      21ddcdc9   
1      e6c5b5cd      c92f3b61      07c540c4      b04e4670      21ddcdc9   
2      6dc710ed      36103458      8efede7f      3412118d           NaN   
3      18231224      52b8680f      1e88c74f      74ef3502           NaN   
4      8ff4b403      01adbab4      1e88c74f      26b3c7a7           NaN   

  categorical20 categorical21 categorical22 categorical23 categorical24  \
0      b1252a9d      07b5194c           NaN      3a171ecb      c5c50484   
1      5840adea      60f6221e           NaN      3a171ecb      43f13e8b   
2           NaN      e587c466      ad3062eb      3a171ecb      3b183c5c   
3           NaN      6b3a5ca6           NaN      3a171ecb      9117a34a   
4           NaN      21c9516a           NaN      32c7478e      b34f3128   

  categorical25 categorical26  
0      e8b83407      9727dd16  
1      e8b83407      731c3655  
2           NaN           NaN  
3           NaN           NaN  
4           NaN           NaN

Now we will verify the dataset shape with the dataset description

In [7]:
len(df.columns)

40

In [8]:
len(answer.columns)

39

In [9]:
df.columns

Index(['label', 'interger1', 'interger2', 'interger3', 'interger4',
       'interger5', 'interger6', 'interger7', 'interger8', 'interger9',
       'interger10', 'interger11', 'interger12', 'interger13', 'categorical1',
       'categorical2', 'categorical3', 'categorical4', 'categorical5',
       'categorical6', 'categorical7', 'categorical8', 'categorical9',
       'categorical10', 'categorical11', 'categorical12', 'categorical13',
       'categorical14', 'categorical15', 'categorical16', 'categorical17',
       'categorical18', 'categorical19', 'categorical20', 'categorical21',
       'categorical22', 'categorical23', 'categorical24', 'categorical25',
       'categorical26'],
      dtype='object')

In [10]:
df.groupby(df.label).interger1.mean().compute()

label
0    2.923491
1    4.771242
Name: interger1, dtype: float64

In [11]:
df.groupby(df.label)['interger1', 'interger2', 'interger3', 'interger4',
       'interger5', 'interger6', 'interger7', 'interger8', 'interger9',
       'interger10', 'interger11', 'interger12', 'interger13'].mean().compute()

interger1   interger2  interger3  interger4     interger5   interger6  \
label                                                                          
0       2.923491   95.639159  23.892997   7.598287  21823.780867  137.242745   
1       4.771242  135.484315  36.351106   6.491826   9210.846171   64.987834   

       interger7  interger8   interger9  interger10  interger11  interger12  \
label                                                                         
0      13.043355  12.787176  103.624690    0.536868    2.251343    0.679853   
1      25.631532  11.733193  113.133942    0.794316    4.093748    1.759554   

       interger13  
label              
0        8.965261  
1        5.963113

## Null verification


Since the dataset is full of nulls we will check them

In [12]:
df.isnull().mean().compute().sort_values()

label            0.000000
categorical11    0.000000
categorical10    0.000000
categorical9     0.000000
categorical8     0.000000
categorical7     0.000000
categorical14    0.000000
categorical5     0.000000
categorical2     0.000000
categorical1     0.000000
categorical15    0.000000
categorical18    0.000000
categorical23    0.000000
interger2        0.000000
categorical17    0.000000
categorical13    0.000000
interger8        0.000497
interger5        0.025809
categorical16    0.034019
categorical12    0.034019
categorical3     0.034019
categorical4     0.034019
categorical21    0.034019
categorical24    0.034019
interger11       0.043256
interger9        0.043256
interger7        0.043256
categorical6     0.120867
interger3        0.214645
interger13       0.216781
interger4        0.216781
interger6        0.223652
categorical19    0.440065
categorical20    0.440065
categorical25    0.440065
categorical26    0.440065
interger10       0.453606
interger1        0.453606
categorical2

## Categories levels

Let's verify how many levels each category has

In [13]:
len(df.categorical11.value_counts().compute())

5683

Variable 11 that has the lower amount of nulls has 5683 categories :O

In [14]:
len(df.categorical10.value_counts().compute())

93145

In [15]:
len(df.categorical9.value_counts().compute())

3

In [16]:
len(df.categorical8.value_counts().compute())

633

In [17]:
len(df.categorical7.value_counts().compute())

12517

In [18]:
len(df.categorical14.value_counts().compute())

27

In [19]:
len(df.categorical5.value_counts().compute())

305

In [20]:
len(df.categorical2.value_counts().compute())

583

In [21]:
len(df.categorical1.value_counts().compute())

1460

In [22]:
len(df.categorical15.value_counts().compute())

14992

In [23]:
len(df.categorical18.value_counts().compute())

5652

In [24]:
len(df.categorical23.value_counts().compute())

15

In [25]:
len(df.categorical17.value_counts().compute())

10

In [26]:
len(df.categorical13.value_counts().compute())

3194

In [27]:
len(df.categorical16.value_counts().compute())

5461305

In [28]:
len(df.categorical12.value_counts().compute())

8351592

In [29]:
len(df.categorical3.value_counts().compute())

10131226

In [30]:
len(df.categorical4.value_counts().compute())

2202607

In [31]:
len(df.categorical21.value_counts().compute())

7046546

In [32]:
len(df.categorical24.value_counts().compute())

286180

In [33]:
len(df.categorical6.value_counts().compute())

23

In [34]:
len(df.categorical19.value_counts().compute())

2172

Integer variables

In [35]:
# plotar o describe das vars

In [36]:
df.interger2.describe().compute()

count    4.584062e+07
mean     1.058484e+02
std      3.914578e+02
min     -3.000000e+00
25%      0.000000e+00
50%      4.000000e+00
75%      4.000000e+01
max      2.576750e+05
dtype: float64

In [37]:
df.interger8.describe().compute()

count    4.581784e+07
mean     1.251704e+01
std      1.668888e+01
min      0.000000e+00
25%      3.000000e+00
50%      8.000000e+00
75%      2.100000e+01
max      6.047000e+03
dtype: float64

In [38]:
df.interger5.describe().compute()

count    4.465750e+07
mean     1.853899e+04
std      6.939460e+04
min      0.000000e+00
25%      9.480000e+02
50%      4.148000e+03
75%      1.119800e+04
max      2.315946e+07
dtype: float64

In [39]:
df.interger11.describe().compute()

count    4.385775e+07
mean     2.732834e+00
std      5.199071e+00
min      0.000000e+00
25%      1.000000e+00
50%      1.000000e+00
75%      3.000000e+00
max      2.310000e+02
dtype: float64

In [40]:
df.interger9.describe().compute()

count    4.385775e+07
mean     1.061098e+02
std      2.202831e+02
min      0.000000e+00
25%      1.200000e+01
50%      4.500000e+01
75%      1.300000e+02
max      2.901900e+04
dtype: float64

In [41]:
df.interger7.describe().compute()

count    4.385775e+07
mean     1.633313e+01
std      6.604976e+01
min      0.000000e+00
25%      1.000000e+00
50%      4.000000e+00
75%      1.300000e+01
max      5.631100e+04
dtype: float64

In [42]:
df.interger3.describe().compute()

count    3.600117e+07
mean     2.691304e+01
std      3.979726e+02
min      0.000000e+00
25%      3.000000e+00
50%      8.000000e+00
75%      2.400000e+01
max      6.553500e+04
dtype: float64

In [43]:
df.interger13.describe().compute()

count    3.590325e+07
mean     8.217461e+00
std      1.621193e+01
min      0.000000e+00
25%      2.000000e+00
50%      5.000000e+00
75%      1.300000e+01
max      7.393000e+03
dtype: float64

In [44]:
df.interger4.describe().compute()

count    3.590325e+07
mean     7.322680e+00
std      8.793231e+00
min      0.000000e+00
25%      2.000000e+00
50%      5.000000e+00
75%      1.100000e+01
max      9.690000e+02
dtype: float64

In [45]:
df.interger6.describe().compute()

count    3.558829e+07
mean     1.160619e+02
std      3.825664e+02
min      0.000000e+00
25%      9.000000e+00
50%      3.700000e+01
75%      1.170000e+02
max      4.310370e+05
dtype: float64

Now we will use a linear classifier with the numerical variables as a baseline for our implementation.

In [47]:
df['label'].describe().compute()

count    4.584062e+07
mean     2.562234e-01
std      4.365466e-01
min      0.000000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      1.000000e+00
max      1.000000e+00
dtype: float64

In [9]:
integer_data = df[['interger1', 'interger2', 'interger3', 'interger4',
       'interger5', 'interger6', 'interger7', 'interger8', 'interger9',
       'interger10', 'interger11', 'interger12', 'interger13']]

In [5]:
from dask.distributed import Client 
client = Client() # start a local Dask client

In [ ]:
from sklearn.externals import joblib
from sklearn.externals.joblib import parallel_backend
from sklearn.ensemble import RandomForestClassifier 
with parallel_backend('dask'):
    # Your normal scikit-learn code here
    model = RandomForestClassifier()
    model.fit(integer_data, df.label)

/home/bahbbc/anaconda3/lib/python3.6/site-packages/sklearn/ensemble/forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)
